File: app/services/blast_service.rb
class BlastService
  def process_results(xml_results)
    hits = []
    xml_results.search("//Hit").each do |hit|
      hit_def = hit.search("Hit_def").text
      hit_accession = hit.search("Hit_accession").text
      hit_len = hit.search("Hit_len").text.to_i
      
      # Extract ID from hit_def (assuming format like "lcl|123 description")
      seq_id = hit_def.split('|')[1]&.split(' ')&.first || hit_accession
      
      # Find the sequence in the database
      sequence = Sequence.find_by(id: seq_id)
      
      # Initialize gene information
      gene_symbol = nil
      original_name = hit_def
      
      if sequence
        # Try to get gene symbol through different approaches
        gene_symbol = find_gene_symbol(sequence, hit_def)
        
        # Get original name if available
        if sequence.respond_to?(:name) && sequence.name.present?
          original_name = sequence.name
        elsif sequence.respond_to?(:human_gene) && sequence.human_gene&.original_name.present?
          original_name = sequence.human_gene.original_name
        end
      end
      
      # Get the first HSP (High-scoring Segment Pair)
      hsp = hit.search("Hit_hsps/Hsp").first
      
      if hsp
        hsp_bit_score = hsp.search("Hsp_bit-score").text.to_f
        hsp_score = hsp.search("Hsp_score").text.to_i
        hsp_evalue = hsp.search("Hsp_evalue").text.to_f
        hsp_query_from = hsp.search("Hsp_query-from").text.to_i
        hsp_query_to = hsp.search("Hsp_query-to").text.to_i
        hsp_hit_from = hsp.search("Hsp_hit-from").text.to_i
        hsp_hit_to = hsp.search("Hsp_hit-to").text.to_i
        hsp_identity = hsp.search("Hsp_identity").text.to_i
        hsp_positive = hsp.search("Hsp_positive").text.to_i
        hsp_gaps = hsp.search("Hsp_gaps").text.to_i
        hsp_align_len = hsp.search("Hsp_align-len").text.to_i
        
        hits << {
          hit_num: hits.count + 1,
          hit_id: seq_id,
          hit_def: hit_def,
          gene_symbol: gene_symbol || '-',
          original_name: original_name,
          hit_len: hit_len,
          hsp_bit_score: hsp_bit_score,
          hsp_score: hsp_score,
          hsp_evalue: hsp_evalue,
          hsp_query_from: hsp_query_from,
          hsp_query_to: hsp_query_to,
          hsp_hit_from: hsp_hit_from,
          hsp_hit_to: hsp_hit_to,
          hsp_identity: hsp_identity,
          hsp_positive: hsp_positive,
          hsp_gaps: hsp_gaps,
          hsp_align_len: hsp_align_len
        }
      end
    end
    hits
  end
  
  private
  
  def find_gene_symbol(sequence, hit_def)
    # Check direct attributes first (no Rails dependencies)
    if has_attribute?(sequence, :human_gene) && sequence.human_gene
      return sequence.human_gene.hgnc_symbol if has_attribute?(sequence.human_gene, :hgnc_symbol) && sequence.human_gene.hgnc_symbol
    end
    
    # Check gene_name attribute
    if has_attribute?(sequence, :gene_name) && !empty?(sequence.gene_name)
      return sequence.gene_name
    end
    
    # Try to find gene symbol by name or id
    name_to_try = nil
    
    # Try sequence name if available
    if has_attribute?(sequence, :name) && !empty?(sequence.name)
      name_to_try = sequence.name
    end
    
    # Then try extracting from hit_def
    if empty?(name_to_try) || name_to_try == hit_def
      name_to_try = extract_gene_name_from_hit_def(hit_def)
    end
    
    # Try sequence ID if it looks like a gene name
    if empty?(name_to_try) && sequence.id.to_s =~ /^[A-Z0-9]+$/i
      name_to_try = sequence.id.to_s
    end
    
    # Try database lookup if we can access database
    if !empty?(name_to_try) && defined?(HumanGene)
      human_gene = simple_find_human_gene(name_to_try)
      if human_gene && has_attribute?(human_gene, :hgnc_symbol) && !empty?(human_gene.hgnc_symbol)
        return human_gene.hgnc_symbol
      end
    end
    
    nil
  end
  
  # Helper methods that avoid Rails dependencies
  
  def has_attribute?(obj, attr)
    obj.respond_to?(attr)
  end
  
  def empty?(value)
    value.nil? || (value.respond_to?(:empty?) && value.empty?)
  end
  
  # Simplified human gene lookup without Rails-specific queries
  def simple_find_human_gene(name)
    return nil if empty?(name)
    
    # Try to find by name however your data access layer allows
    # This is a simplified placeholder - implement based on your data access method
    if defined?(HumanGene) && HumanGene.respond_to?(:find_by_name)
      HumanGene.find_by_name(name)
    elsif defined?(HumanGene) && HumanGene.respond_to?(:where)
      # Minimal query if ActiveRecord-like interface is available
      HumanGene.where(original_name: name).first
    else
      nil
    end
  end
  
  def extract_gene_name_from_hit_def(hit_def)
    # Attempt to extract gene names like MGC2752 from hit_def
    gene_patterns = [
      /\b(MGC\d+)\b/i,                    # Match MGC followed by numbers
      /\b([A-Z0-9]{2,})\s+protein\b/i,    # Match gene name followed by "protein"
      /protein\s+([A-Z0-9]{2,})\b/i,      # Match "protein" followed by gene name
      /\b([A-Z]{2,}[0-9]{1,})\b/i         # Match letters followed by numbers
    ]
    
    gene_patterns.each do |pattern|
      match = hit_def.match(pattern)
      return match[1] if match
    end
    
    nil
  end
end

File: services/blast_service.py
import re
from bs4 import BeautifulSoup
from models.sequence import Sequence
from models.human_gene import HumanGene

class BlastService:
    """Service for processing BLAST search results."""
    
    def process_results(self, xml_results):
        """Process BLAST XML results into a structured format."""
        hits = []
        soup = BeautifulSoup(xml_results, 'xml')
        
        for hit_num, hit in enumerate(soup.find_all('Hit'), 1):
            hit_def = hit.find('Hit_def').text
            hit_accession = hit.find('Hit_accession').text
            hit_len = int(hit.find('Hit_len').text)
            
            # Extract ID from hit_def (assuming format like "lcl|123 description")
            seq_id = hit_accession
            if '|' in hit_def:
                parts = hit_def.split('|')
                if len(parts) > 1 and parts[1]:
                    id_part = parts[1].split(' ')[0]
                    if id_part:
                        seq_id = id_part
            
            # Find the sequence in the database
            sequence = Sequence.find_by_id(seq_id)
            
            # Initialize gene information
            gene_symbol = None
            original_name = hit_def
            
            if sequence:
                # Try to get gene symbol through different approaches
                gene_symbol = self.find_gene_symbol(sequence, hit_def)
                
                # Get original name if available
                if hasattr(sequence, 'name') and sequence.name:
                    original_name = sequence.name
                elif hasattr(sequence, 'human_gene') and sequence.human_gene and hasattr(sequence.human_gene, 'original_name'):
                    original_name = sequence.human_gene.original_name
            
            # Get the first HSP (High-scoring Segment Pair)
            hsp = hit.find('Hsp')
            
            if hsp:
                hit_data = {
                    'hit_num': hit_num,
                    'hit_id': seq_id,
                    'hit_def': hit_def,
                    'gene_symbol': gene_symbol or '-',
                    'original_name': original_name,
                    'hit_len': hit_len,
                    'hsp_bit_score': float(hsp.find('Hsp_bit-score').text),
                    'hsp_score': int(hsp.find('Hsp_score').text),
                    'hsp_evalue': float(hsp.find('Hsp_evalue').text),
                    'hsp_query_from': int(hsp.find('Hsp_query-from').text),
                    'hsp_query_to': int(hsp.find('Hsp_query-to').text),
                    'hsp_hit_from': int(hsp.find('Hsp_hit-from').text),
                    'hsp_hit_to': int(hsp.find('Hsp_hit-to').text),
                    'hsp_identity': int(hsp.find('Hsp_identity').text),
                    'hsp_positive': int(hsp.find('Hsp_positive').text),
                    'hsp_gaps': int(hsp.find('Hsp_gaps').text),
                    'hsp_align_len': int(hsp.find('Hsp_align-len').text)
                }
                hits.append(hit_data)
        
        return hits
    
    def find_gene_symbol(self, sequence, hit_def):
        """Find gene symbol using various approaches."""
        # Check if sequence has a direct association with human_gene
        if hasattr(sequence, 'human_gene') and sequence.human_gene:
            if hasattr(sequence.human_gene, 'hgnc_symbol') and sequence.human_gene.hgnc_symbol:
                return sequence.human_gene.hgnc_symbol
        
        # Check if sequence has a gene_name attribute
        if hasattr(sequence, 'gene_name') and sequence.gene_name:
            return sequence.gene_name
        
        # Try to find gene symbol by name or id
        name_to_try = None
        
        # First try using sequence name if available
        if hasattr(sequence, 'name') and sequence.name:
            name_to_try = sequence.name
        
        # Then try extracting from hit_def
        if not name_to_try or name_to_try == hit_def:
            name_to_try = self.extract_gene_name_from_hit_def(hit_def)
        
        # Next try the sequence ID itself if it looks like a gene name
        if not name_to_try and re.match(r'^[A-Za-z0-9]+$', str(sequence.id)):
            name_to_try = str(sequence.id)
        
        # Try to find a human gene with that original name
        if name_to_try:
            human_gene = HumanGene.find_by_name(name_to_try)
            if human_gene and hasattr(human_gene, 'hgnc_symbol') and human_gene.hgnc_symbol:
                return human_gene.hgnc_symbol
        
        return None
    
    def extract_gene_name_from_hit_def(self, hit_def):
        """Extract gene names from hit definition using regex patterns."""
        gene_patterns = [
            r'\b(MGC\d+)\b',                     # Match MGC followed by numbers
            r'\b([A-Z0-9]{2,})\s+protein\b',     # Match gene name followed by "protein" 
            r'protein\s+([A-Z0-9]{2,})\b',       # Match "protein" followed by gene name
            r'\b([A-Z]{2,}[0-9]{1,})\b'          # Match letters followed by numbers
        ]
        
        for pattern in gene_patterns:
            match = re.search(pattern, hit_def, re.IGNORECASE)
            if match:
                return match.group(1)
        
        return None
